Framework/Nexus/inc/MantidNexus/NexusIOHelper.h

```diff
@@ -0,0 +1,162 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2007 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+#ifndef NEXUSIOHELPER_H
+#define NEXUSIOHELPER_H
+
+#include "MantidIndexing/DllConfig.h"
+#include <algorithm>
+#include <boost/any.hpp>
+#include <nexus/NeXusFile.hpp>
+#include <numeric>
+#include <string>
+
+namespace Mantid {
+namespace NeXus {
+namespace NeXusIOHelper {
+
+namespace {
+
+/// Macro to run a function depending on the type of data in the Nexus file
+#define RUN_NEXUSIOHELPER_FUNCTION(type, func_name, ...)                       \
+  switch (type) {                                                              \
+  case ::NeXus::FLOAT32:                                                       \
+    return func_name<T, float>(__VA_ARGS__);                                   \
+  case ::NeXus::FLOAT64:                                                       \
+    return func_name<T, double>(__VA_ARGS__);                                  \
+  case ::NeXus::INT8:                                                          \
+    return func_name<T, int8_t>(__VA_ARGS__);                                  \
+  case ::NeXus::UINT8:                                                         \
+    return func_name<T, uint8_t>(__VA_ARGS__);                                 \
+  case ::NeXus::INT16:                                                         \
+    return func_name<T, int16_t>(__VA_ARGS__);                                 \
+  case ::NeXus::UINT16:                                                        \
+    return func_name<T, uint16_t>(__VA_ARGS__);                                \
+  case ::NeXus::INT32:                                                         \
+    return func_name<T, int32_t>(__VA_ARGS__);                                 \
+  case ::NeXus::UINT32:                                                        \
+    return func_name<T, uint32_t>(__VA_ARGS__);                                \
+  case ::NeXus::INT64:                                                         \
+    return func_name<T, int64_t>(__VA_ARGS__);                                 \
+  case ::NeXus::UINT64:                                                        \
+    return func_name<T, uint64_t>(__VA_ARGS__);                                \
+  default:                                                                     \
+    throw std::runtime_error("Unknown type in Nexus file");                    \
+  }
+
+std::pair<::NeXus::Info, bool> checkIfOpenAndGetInfo(::NeXus::File &file,
```

Comment:
Without looking at the definition, `checkIf` prefix might imply that there would be exceptions thrown. Maybe we should drop the `checkifOpen` prefix, since you open it if closed as an implementation detail?